core/ir/riscv64/codec.c

```diff
@@ -0,0 +1,1066 @@
+    instr_set_src(out, idx, opnd);
+    return true;
+}
+
+static bool
+decode_csr_opnd(dcontext_t *dc, uint32_t inst, int op_sz, byte *pc, byte *orig_pc,
+                int idx, instr_t *out)
+{
+    /* FIXME i#3544: Should CSRs be as DR_REG_* or rather as hex defines? Their
+     * set is extensible by platform implementers and various extensions, so
+     * for now let's leave it as an int.
+     */
+    int32_t imm = GET_FIELD(inst, 31, 20);
+    opnd_t opnd = opnd_create_immed_int(imm, op_sz);
+    instr_set_src(out, idx, opnd);
+    return true;
+}
+
+static bool
+decode_rm_opnd(dcontext_t *dc, uint32_t inst, int op_sz, byte *pc, byte *orig_pc, int idx,
+               instr_t *out)
+{
+    int32_t imm = GET_FIELD(inst, 14, 12);
+    opnd_t opnd = opnd_create_immed_int(imm, op_sz);
+    instr_set_src(out, idx, opnd);
+    return true;
+}
+
+static bool
+decode_shamt_opnd(dcontext_t *dc, uint32_t inst, int op_sz, byte *pc, byte *orig_pc,
+                  int idx, instr_t *out)
+{
+    int32_t imm = GET_FIELD(inst, 25, 20);
+    opnd_t opnd = opnd_create_immed_int(imm, op_sz);
+    instr_set_src(out, idx, opnd);
+    return true;
+}
+
+static bool
+decode_shamt5_opnd(dcontext_t *dc, uint32_t inst, int op_sz, byte *pc, byte *orig_pc,
+                   int idx, instr_t *out)
+{
+    int32_t imm = GET_FIELD(inst, 24, 20);
+    opnd_t opnd = opnd_create_immed_int(imm, op_sz);
+    instr_set_src(out, idx, opnd);
+    return true;
+}
+
+static bool
+decode_shamt6_opnd(dcontext_t *dc, uint32_t inst, int op_sz, byte *pc, byte *orig_pc,
+                   int idx, instr_t *out)
+{
+    /* shamt6 >= 64 only makes sense on RV128 but let user take care of it. */
+    int32_t imm = GET_FIELD(inst, 26, 20);
+    opnd_t opnd = opnd_create_immed_int(imm, op_sz);
+    instr_set_src(out, idx, opnd);
+    return true;
+}
+
+static bool
+decode_i_imm_opnd(dcontext_t *dc, uint32_t inst, int op_sz, byte *pc, byte *orig_pc,
+                  int idx, instr_t *out)
+{
+    int32_t imm = SIGN_EXTEND(GET_FIELD(inst, 31, 20), 12);
+    opnd_t opnd = opnd_create_immed_int(imm, op_sz);
+    instr_set_src(out, idx, opnd);
+    return true;
+}
+
+static bool
+decode_s_imm_opnd(dcontext_t *dc, uint32_t inst, int op_sz, byte *pc, byte *orig_pc,
+                  int idx, instr_t *out)
+{
+    int32_t imm = (GET_FIELD(inst, 31, 25) << 5) | (GET_FIELD(inst, 11, 7));
+    imm = SIGN_EXTEND(imm, 12);
+    opnd_t opnd = opnd_create_immed_int(imm, op_sz);
+    instr_set_src(out, idx, opnd);
+    return true;
+}
+
+static bool
+decode_b_imm_opnd(dcontext_t *dc, uint32_t inst, int op_sz, byte *pc, byte *orig_pc,
+                  int idx, instr_t *out)
+{
+    int32_t imm = BIT(inst, 31) << 12;
+    imm |= BIT(inst, 7) << 11;
+    imm |= GET_FIELD(inst, 30, 25) << 5;
+    imm |= GET_FIELD(inst, 11, 8) << 1;
+    imm = SIGN_EXTEND(imm, 13);
+    /* FIXME i#3544: Should PC-relative jumps targets be encoded as mem_instr or
+     * rather rel_addr?
```

Comment:
> You mean an indirect jump whose target is stored in memory at a PC-relative address? Does RISC-V have such instructions? I thought its indirect jumps stored their addresses in registers only?

No, I meant instructions such as "JAL" (jump and link) or "BEQZ" (branch if equal zero). Both compute the jump target by adding the immediate encoded in the instruction to the current value of the PC register. So I thought that I should decode the "immediate" operand as either `opnd_create_rel_addr()` or `opnd_create_mem_instr()`.

Though looking at ARM (<=v7) I should just use `opnd_create_base_disp()` with `DR_REG_PC` as base?

Reply:
I see.  Today all arches (x86, aarch64, aarch32) use `opnd_create_pc` to store a direct jump/branch's target as an absolute PC, resolved using the current PC at decode time.  It is implicitly a PC-relative type of operand and follows the general rule of having absolute resolved addresses in operands for easy examination.  When encoding an unchanged instruction there is supposed to be support to auto-magically re-relativize to still point at the same absolute address (instead of maintaining the relative offset from the new encoding position): though there are still some holes there (#4016) as it does not affect core operation.

Reply:
Done.